File: packages/fluent-cards/src/builders/AdaptiveCardBuilder.ts
import type {
  AdaptiveCard,
  AdaptiveAction,
  AdaptiveElement,
  AuthenticationConfiguration,
  BackgroundImage,
  CardMetadata,
  Column,
  RefreshConfiguration,
} from '../models.js';
import { TextBlockBuilder } from './TextBlockBuilder.js';
import { ImageBuilder } from './ImageBuilder.js';
import { ContainerBuilder } from './ContainerBuilder.js';
import { ColumnSetBuilder } from './ColumnSetBuilder.js';
import { FactSetBuilder } from './FactSetBuilder.js';
import { RichTextBlockBuilder } from './RichTextBlockBuilder.js';
import { ActionSetBuilder } from './ActionSetBuilder.js';
import { MediaBuilder } from './MediaBuilder.js';
import { ImageSetBuilder } from './ImageSetBuilder.js';
import { TableBuilder } from './TableBuilder.js';
import { ActionBuilder } from './ActionBuilder.js';
import { RefreshBuilder } from './RefreshBuilder.js';
import { AuthenticationBuilder } from './AuthenticationBuilder.js';
import { InputTextBuilder } from './inputs/InputTextBuilder.js';
import { InputNumberBuilder } from './inputs/InputNumberBuilder.js';
import { InputDateBuilder } from './inputs/InputDateBuilder.js';
import { InputTimeBuilder } from './inputs/InputTimeBuilder.js';
import { InputToggleBuilder } from './inputs/InputToggleBuilder.js';
import { InputChoiceSetBuilder } from './inputs/InputChoiceSetBuilder.js';

/** Fluent builder for creating {@link AdaptiveCard} instances. */
export class AdaptiveCardBuilder {
  private readonly card: AdaptiveCard = {
    type: 'AdaptiveCard',
    version: '1.5',
    '$schema': 'http://adaptivecards.io/schemas/adaptive-card.json',
  };

  static create(): AdaptiveCardBuilder {
    return new AdaptiveCardBuilder();
  }

  withVersion(version: string): this {
    this.card.version = version;
    return this;
  }

  withSchema(schema: string | undefined): this {
    this.card['$schema'] = schema;
    return this;
  }

  withMetadata(webUrl: string): this {
    this.card.metadata = { webUrl };
    return this;
  }

  // ─── Body elements ────────────────────────────────────────────────────────

  addTextBlock(configure: (b: TextBlockBuilder) => void): this {
    const b = new TextBlockBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addImage(configure: (b: ImageBuilder) => void): this {
    const b = new ImageBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addContainer(configure: (b: ContainerBuilder) => void): this {
    const b = new ContainerBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addColumnSet(configure: (b: ColumnSetBuilder) => void): this {
    const b = new ColumnSetBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addFactSet(configure: (b: FactSetBuilder) => void): this {
    const b = new FactSetBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addRichTextBlock(configure: (b: RichTextBlockBuilder) => void): this {
    const b = new RichTextBlockBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addActionSet(configure: (b: ActionSetBuilder) => void): this {
    const b = new ActionSetBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addMedia(configure: (b: MediaBuilder) => void): this {
    const b = new MediaBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addImageSet(configure: (b: ImageSetBuilder) => void): this {
    const b = new ImageSetBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addTable(configure: (b: TableBuilder) => void): this {
    const b = new TableBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  // ─── Input elements ───────────────────────────────────────────────────────

  addInputText(configure: (b: InputTextBuilder) => void): this {
    const b = new InputTextBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addInputNumber(configure: (b: InputNumberBuilder) => void): this {
    const b = new InputNumberBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addInputDate(configure: (b: InputDateBuilder) => void): this {
    const b = new InputDateBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addInputTime(configure: (b: InputTimeBuilder) => void): this {
    const b = new InputTimeBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addInputToggle(configure: (b: InputToggleBuilder) => void): this {
    const b = new InputToggleBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  addInputChoiceSet(configure: (b: InputChoiceSetBuilder) => void): this {
    const b = new InputChoiceSetBuilder();
    configure(b);
    this.pushBody(b.build());
    return this;
  }

  // ─── Actions ─────────────────────────────────────────────────────────────

  addAction(configure: (b: ActionBuilder) => void): this {
    const b = new ActionBuilder();
    configure(b);
    (this.card.actions ??= []).push(b.build());
    return this;
  }

  // ─── Advanced configuration ───────────────────────────────────────────────

  withRefresh(configure: (b: RefreshBuilder) => void): this {
    const b = new RefreshBuilder();
    configure(b);
    this.card.refresh = b.build();
    return this;
  }

  withAuthentication(configure: (b: AuthenticationBuilder) => void): this {
    const b = new AuthenticationBuilder();
    configure(b);
    this.card.authentication = b.build();
    return this;
  }

  build(): AdaptiveCard {
    return this.card;
  }

  private pushBody(element: AdaptiveElement): void {
    (this.card.body ??= []).push(element);
  }
}

File: java/src/main/java/io/fluentcards/AdaptiveCardBuilder.java
package io.fluentcards;

import java.util.*;
import java.util.function.Consumer;

public class AdaptiveCardBuilder {
    static final Map<String, String> SCHEMA_URLS;
    static final Set<String> KNOWN_VERSIONS = Set.of("1.0", "1.1", "1.2", "1.3", "1.4", "1.5", "1.6");

    static {
        Map<String, String> urls = new LinkedHashMap<>();
        urls.put("1.0", "https://adaptivecards.io/schemas/1.0.0/adaptive-card.json");
        urls.put("1.1", "https://adaptivecards.io/schemas/1.1.0/adaptive-card.json");
        urls.put("1.2", "https://adaptivecards.io/schemas/1.2.0/adaptive-card.json");
        urls.put("1.3", "https://adaptivecards.io/schemas/1.3.0/adaptive-card.json");
        urls.put("1.4", "https://adaptivecards.io/schemas/1.4.0/adaptive-card.json");
        urls.put("1.5", "https://adaptivecards.io/schemas/1.5.0/adaptive-card.json");
        urls.put("1.6", "https://adaptivecards.io/schemas/1.6.0/adaptive-card.json");
        SCHEMA_URLS = Collections.unmodifiableMap(urls);
    }

    private final LinkedHashMap<String, Object> data;

    public AdaptiveCardBuilder() {
        data = new LinkedHashMap<>();
        data.put("type", "AdaptiveCard");
        data.put("version", "1.5");
        data.put("$schema", SCHEMA_URLS.get("1.5"));
    }

    public static AdaptiveCardBuilder create() {
        return new AdaptiveCardBuilder();
    }

    public AdaptiveCardBuilder withVersion(String version) {
        data.put("version", version);
        String schema = SCHEMA_URLS.get(version);
        if (schema != null) {
            data.put("$schema", schema);
        }
        return this;
    }

    public AdaptiveCardBuilder withSchema(String schema) {
        data.put("$schema", schema);
        return this;
    }

    public AdaptiveCardBuilder withFallbackText(String fallbackText) {
        data.put("fallbackText", fallbackText);
        return this;
    }

    public AdaptiveCardBuilder withSpeak(String speak) {
        data.put("speak", speak);
        return this;
    }

    public AdaptiveCardBuilder withLang(String lang) {
        data.put("lang", lang);
        return this;
    }

    public AdaptiveCardBuilder withRtl(boolean rtl) {
        data.put("rtl", rtl);
        return this;
    }

    public AdaptiveCardBuilder withMinHeight(String minHeight) {
        data.put("minHeight", minHeight);
        return this;
    }

    public AdaptiveCardBuilder withVerticalContentAlignment(VerticalAlignment alignment) {
        data.put("verticalContentAlignment", alignment.getValue());
        return this;
    }

    public AdaptiveCardBuilder withBackgroundImage(Consumer<BackgroundImageBuilder> configure) {
        BackgroundImageBuilder builder = new BackgroundImageBuilder();
        configure.accept(builder);
        data.put("backgroundImage", builder.build());
        return this;
    }

    public AdaptiveCardBuilder withSelectAction(Consumer<ActionBuilder> configure) {
        ActionBuilder builder = new ActionBuilder();
        configure.accept(builder);
        data.put("selectAction", builder.build());
        return this;
    }

    public AdaptiveCardBuilder withMetadata(String webUrl) {
        LinkedHashMap<String, Object> metadata = new LinkedHashMap<>();
        metadata.put("webUrl", webUrl);
        data.put("metadata", metadata);
        return this;
    }

    // Body element adders

    public AdaptiveCardBuilder addTextBlock(Consumer<TextBlockBuilder> configure) {
        TextBlockBuilder builder = new TextBlockBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addImage(Consumer<ImageBuilder> configure) {
        ImageBuilder builder = new ImageBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addContainer(Consumer<ContainerBuilder> configure) {
        ContainerBuilder builder = new ContainerBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addColumnSet(Consumer<ColumnSetBuilder> configure) {
        ColumnSetBuilder builder = new ColumnSetBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addFactSet(Consumer<FactSetBuilder> configure) {
        FactSetBuilder builder = new FactSetBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addRichTextBlock(Consumer<RichTextBlockBuilder> configure) {
        RichTextBlockBuilder builder = new RichTextBlockBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addActionSet(Consumer<ActionSetBuilder> configure) {
        ActionSetBuilder builder = new ActionSetBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addMedia(Consumer<MediaBuilder> configure) {
        MediaBuilder builder = new MediaBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addImageSet(Consumer<ImageSetBuilder> configure) {
        ImageSetBuilder builder = new ImageSetBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addTable(Consumer<TableBuilder> configure) {
        TableBuilder builder = new TableBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addElement(Map<String, Object> element) {
        pushBody(element);
        return this;
    }

    // Input adders

    public AdaptiveCardBuilder addInputText(Consumer<InputTextBuilder> configure) {
        InputTextBuilder builder = new InputTextBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addInputNumber(Consumer<InputNumberBuilder> configure) {
        InputNumberBuilder builder = new InputNumberBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addInputDate(Consumer<InputDateBuilder> configure) {
        InputDateBuilder builder = new InputDateBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addInputTime(Consumer<InputTimeBuilder> configure) {
        InputTimeBuilder builder = new InputTimeBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addInputToggle(Consumer<InputToggleBuilder> configure) {
        InputToggleBuilder builder = new InputToggleBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    public AdaptiveCardBuilder addInputChoiceSet(Consumer<InputChoiceSetBuilder> configure) {
        InputChoiceSetBuilder builder = new InputChoiceSetBuilder();
        configure.accept(builder);
        pushBody(builder.build());
        return this;
    }

    // Actions

    @SuppressWarnings("unchecked")
    public AdaptiveCardBuilder addAction(Consumer<ActionBuilder> configure) {
        ActionBuilder builder = new ActionBuilder();
        configure.accept(builder);
        List<Object> actions = (List<Object>) data.get("actions");
        if (actions == null) {
            actions = new ArrayList<>();
            data.put("actions", actions);
        }
        actions.add(builder.build());
        return this;
    }

    // Refresh & Authentication

    public AdaptiveCardBuilder withRefresh(Consumer<RefreshBuilder> configure) {
        RefreshBuilder builder = new RefreshBuilder();
        configure.accept(builder);
        data.put("refresh", builder.build());
        return this;
    }

    public AdaptiveCardBuilder withAuthentication(Consumer<AuthenticationBuilder> configure) {
        AuthenticationBuilder builder = new AuthenticationBuilder();
        configure.accept(builder);
        data.put("authentication", builder.build());
        return this;
    }

    public Map<String, Object> build() {
        return data;
    }

    @SuppressWarnings("unchecked")
    private void pushBody(Map<String, Object> element) {
        List<Object> body = (List<Object>) data.get("body");
        if (body == null) {
            body = new ArrayList<>();
            data.put("body", body);
        }
        body.add(element);
    }
}
